File: fw/Zsmt_mark/mcc_generated_files/X2CCode/Library/MCHP_ZSMT/Conversion/C/src/ConvFnc_InitPosDetect_Float32.c
/*
 * Copyright (c) 2018, Linz Center of Mechatronics GmbH (LCM) http://www.lcm.at/
 * All rights reserved.
 */
/*
 * This file is part of X2C. http://x2c.lcm.at/
 * $LastChangedRevision: 2108 $
 * $LastChangedDate:: 2020-12-07 19:09:57 +0100#$
 */
/* USERCODE-BEGIN:Description                                                                                         */
/* Description: */
/* USERCODE-END:Description                                                                                           */
#include "ConvFnc_InitPosDetect_Float32.h"

/* USERCODE-BEGIN:PreProcessor                                                                                        */
/* USERCODE-END:PreProcessor                                                                                          */

extern uint8 MaskParamBuffer[];
extern uint8 ImplParamBuffer[];
extern uint8 SaveFncDataBuffer[];

/** Implementation parameter data structure */
typedef struct {
    float32 u_pulse;
    int16 t_pulse;
    int16 t_period;
} INITPOSDETECT_FLOAT32_IMPL_PARAM;

/** Private prototypes */
static uint8 convertM2I(const INITPOSDETECT_FLOAT32_MASK_PARAM *maskParam, INITPOSDETECT_FLOAT32_IMPL_PARAM *impParam, float_CoT ts);

/**
 * @brief Converts Mask parameters to Implementation parameters.
 *
 * @param[in] maskParam Mask parameters
 * @param[out] impParam Implementation parameters
 * @param[in] ts Block sample time
 *
 * @return Error: zero on success, not zero in case of conversion error
 */
static uint8 convertM2I(const INITPOSDETECT_FLOAT32_MASK_PARAM *maskParam, INITPOSDETECT_FLOAT32_IMPL_PARAM *impParam, float_CoT ts)
{
    uint8 error = (uint8)0;
/* USERCODE-BEGIN:Conversion                                                                                          */
    float_CoT ampPhasor, deltaT, nPulse, nPause, nPeriod;
    
    /** - get mask parameter value */
    ampPhasor = (float_CoT)maskParam->U_pulse;  /* get amplitude of voltage phasor */
    nPulse    = (float_CoT)maskParam->T_pulse;  /* pulse duration in samples */
    deltaT    = (float_CoT)maskParam->T_pause;  /* pause time */

    /** - parameter validation */
    if (nPulse < 1)
    {
        error += 1; /* Pulse duration must be at least one sample! */
    }
    if (deltaT < 0)
    {
        error += 2; /* Pause time must not be negative! */
    }

    /** -  calculate implementation parameter value */
    if (error == 0)
    {
        if (nPulse > 32767)
        {
            nPulse = 32767;
        } 
        nPause = deltaT / ((float_CoT)1000 * ts);      
        nPeriod = (2 * nPulse) + nPause;
        if (nPeriod > 32767)
        {
            nPeriod = 32767;
        }         
        impParam->u_pulse = (float32)ampPhasor; /* voltage phasor-amplitude */
        impParam->t_pulse = (int16)nPulse;                          /* pulse time in updates */
        impParam->t_period = (int16)nPeriod;                        /* period in updates */
    }
/* USERCODE-END:Conversion                                                                                            */
    return (error);
}

/**
 * @brief Load block mask parameter data.
 *
 * @param[in] maskParam Mask parameter data structure
 * @param[out] data Data
 * @param[out] dataLen Data length
 * @param[in] maxSize Maximum Service data length
 *
 * @return Error: zero on success, not zero in case of load- or conversion error
 */
uint8 InitPosDetect_Float32_LoadMP(const INITPOSDETECT_FLOAT32_MASK_PARAM *maskParam, uint8 data[], uint16* dataLen, uint16 maxSize)
{
    uint8 error = (uint8)0;
    if ((uint16)40 > maxSize)
    {
        error = (uint8)1;
    }
    else
    {
        data[0] = (uint8)((*(uint64*)&(maskParam->U_pulse)) & 0x00000000000000FF);
        data[1] = (uint8)((*(uint64*)&(maskParam->U_pulse) >> 8) & 0x00000000000000FF);
        data[2] = (uint8)((*(uint64*)&(maskParam->U_pulse) >> 16) & 0x00000000000000FF);
        data[3] = (uint8)((*(uint64*)&(maskParam->U_pulse) >> 24) & 0x00000000000000FF);
        data[4] = (uint8)((*(uint64*)&(maskParam->U_pulse) >> 32) & 0x00000000000000FF);
        data[5] = (uint8)((*(uint64*)&(maskParam->U_pulse) >> 40) & 0x00000000000000FF);
        data[6] = (uint8)((*(uint64*)&(maskParam->U_pulse) >> 48) & 0x00000000000000FF);
        data[7] = (uint8)((*(uint64*)&(maskParam->U_pulse) >> 56) & 0x00000000000000FF);
        data[8] = (uint8)((*(uint64*)&(maskParam->T_pulse)) & 0x00000000000000FF);
        data[9] = (uint8)((*(uint64*)&(maskParam->T_pulse) >> 8) & 0x00000000000000FF);
        data[10] = (uint8)((*(uint64*)&(maskParam->T_pulse) >> 16) & 0x00000000000000FF);
        data[11] = (uint8)((*(uint64*)&(maskParam->T_pulse) >> 24) & 0x00000000000000FF);
        data[12] = (uint8)((*(uint64*)&(maskParam->T_pulse) >> 32) & 0x00000000000000FF);
        data[13] = (uint8)((*(uint64*)&(maskParam->T_pulse) >> 40) & 0x00000000000000FF);
        data[14] = (uint8)((*(uint64*)&(maskParam->T_pulse) >> 48) & 0x00000000000000FF);
        data[15] = (uint8)((*(uint64*)&(maskParam->T_pulse) >> 56) & 0x00000000000000FF);
        data[16] = (uint8)((*(uint64*)&(maskParam->T_pause)) & 0x00000000000000FF);
        data[17] = (uint8)((*(uint64*)&(maskParam->T_pause) >> 8) & 0x00000000000000FF);
        data[18] = (uint8)((*(uint64*)&(maskParam->T_pause) >> 16) & 0x00000000000000FF);
        data[19] = (uint8)((*(uint64*)&(maskParam->T_pause) >> 24) & 0x00000000000000FF);
        data[20] = (uint8)((*(uint64*)&(maskParam->T_pause) >> 32) & 0x00000000000000FF);
        data[21] = (uint8)((*(uint64*)&(maskParam->T_pause) >> 40) & 0x00000000000000FF);
        data[22] = (uint8)((*(uint64*)&(maskParam->T_pause) >> 48) & 0x00000000000000FF);
        data[23] = (uint8)((*(uint64*)&(maskParam->T_pause) >> 56) & 0x00000000000000FF);
        data[24] = (uint8)((*(uint64*)&(maskParam->U_max)) & 0x00000000000000FF);
        data[25] = (uint8)((*(uint64*)&(maskParam->U_max) >> 8) & 0x00000000000000FF);
        data[26] = (uint8)((*(uint64*)&(maskParam->U_max) >> 16) & 0x00000000000000FF);
        data[27] = (uint8)((*(uint64*)&(maskParam->U_max) >> 24) & 0x00000000000000FF);
        data[28] = (uint8)((*(uint64*)&(maskParam->U_max) >> 32) & 0x00000000000000FF);
        data[29] = (uint8)((*(uint64*)&(maskParam->U_max) >> 40) & 0x00000000000000FF);
        data[30] = (uint8)((*(uint64*)&(maskParam->U_max) >> 48) & 0x00000000000000FF);
        data[31] = (uint8)((*(uint64*)&(maskParam->U_max) >> 56) & 0x00000000000000FF);
        data[32] = (uint8)((*(uint64*)&(maskParam->ts_fact)) & 0x00000000000000FF);
        data[33] = (uint8)((*(uint64*)&(maskParam->ts_fact) >> 8) & 0x00000000000000FF);
        data[34] = (uint8)((*(uint64*)&(maskParam->ts_fact) >> 16) & 0x00000000000000FF);
        data[35] = (uint8)((*(uint64*)&(maskParam->ts_fact) >> 24) & 0x00000000000000FF);
        data[36] = (uint8)((*(uint64*)&(maskParam->ts_fact) >> 32) & 0x00000000000000FF);
        data[37] = (uint8)((*(uint64*)&(maskParam->ts_fact) >> 40) & 0x00000000000000FF);
        data[38] = (uint8)((*(uint64*)&(maskParam->ts_fact) >> 48) & 0x00000000000000FF);
        data[39] = (uint8)((*(uint64*)&(maskParam->ts_fact) >> 56) & 0x00000000000000FF);
        *dataLen = (uint16)40;
    }
    return (error);
}

/**
 * @brief Save block mask parameter data.
 *
 * @param[in] block Pointer to block structure
 * @param[out] maskParam Mask parameter data structure
 * @param[in] data Data
 * @param dataLen Length of mask parameter data stream
 *
 * @return Error: zero on success, not zero in case of save- or conversion error
 */
uint8 InitPosDetect_Float32_SaveMP(INITPOSDETECT_FLOAT32 *block, INITPOSDETECT_FLOAT32_MASK_PARAM *maskParam, const uint8 data[], uint16 dataLen)
{
    uint8 error = (uint8)0;

    if (dataLen != (uint16)40)
    {
        error = (uint8)1;
    }
    else
    {
        uint64 tmp64;
        /** - Cache old mask parameters */
        InitPosDetect_Float32_BackupMP(maskParam);

        /** - Save new mask parameters */
        tmp64 = (uint64)data[0] + \
            ((uint64)data[1] << 8) + ((uint64)data[2] << 16) + \
            ((uint64)data[3] << 24) + ((uint64)data[4] << 32) + \
            ((uint64)data[5] << 40) + ((uint64)data[6] << 48) + \
            ((uint64)data[7] << 56);
        maskParam->U_pulse = (float64)(*(float64*)&tmp64);
        tmp64 = (uint64)data[8] + \
            ((uint64)data[9] << 8) + ((uint64)data[10] << 16) + \
            ((uint64)data[11] << 24) + ((uint64)data[12] << 32) + \
            ((uint64)data[13] << 40) + ((uint64)data[14] << 48) + \
            ((uint64)data[15] << 56);
        maskParam->T_pulse = (float64)(*(float64*)&tmp64);
        tmp64 = (uint64)data[16] + \
            ((uint64)data[17] << 8) + ((uint64)data[18] << 16) + \
            ((uint64)data[19] << 24) + ((uint64)data[20] << 32) + \
            ((uint64)data[21] << 40) + ((uint64)data[22] << 48) + \
            ((uint64)data[23] << 56);
        maskParam->T_pause = (float64)(*(float64*)&tmp64);
        tmp64 = (uint64)data[24] + \
            ((uint64)data[25] << 8) + ((uint64)data[26] << 16) + \
            ((uint64)data[27] << 24) + ((uint64)data[28] << 32) + \
            ((uint64)data[29] << 40) + ((uint64)data[30] << 48) + \
            ((uint64)data[31] << 56);
        maskParam->U_max = (float64)(*(float64*)&tmp64);
        tmp64 = (uint64)data[32] + \
            ((uint64)data[33] << 8) + ((uint64)data[34] << 16) + \
            ((uint64)data[35] << 24) + ((uint64)data[36] << 32) + \
            ((uint64)data[37] << 40) + ((uint64)data[38] << 48) + \
            ((uint64)data[39] << 56);
        maskParam->ts_fact = (float64)(*(float64*)&tmp64);

        error = InitPosDetect_Float32_ConvertMP(block, maskParam);

        /** - Check for errors during execution of save function */
        if (error != 0)
        {
            /* Restore old mask parameter */
            InitPosDetect_Float32_RestoreMP(maskParam);
        }

    }
    return (error);
}

/**
 * @brief Initializes Implementation parameters from Mask parameters.
 *
 * @param[out] block Block
 * @param[in] maskParam Mask parameters
 *
 * @return Error: zero on success, not zero in case of initialization error
 */
uint8 InitPosDetect_Float32_InitMP(INITPOSDETECT_FLOAT32 *block, const INITPOSDETECT_FLOAT32_MASK_PARAM *maskParam)
{
    uint8 error = (uint8)0;
    float_CoT block_ts;

    INITPOSDETECT_FLOAT32_IMPL_PARAM *implParam = (INITPOSDETECT_FLOAT32_IMPL_PARAM*)ImplParamBuffer;

    /** - Get Block Sample Time */
    block_ts = maskParam->ts_fact * getModelSampleTime();
    /** - Convert mask parameters to implementation parameters */
    error = convertM2I(maskParam, implParam, block_ts);

    /** - Prepare implementation parameters for save function */
    SaveFncDataBuffer[0] = (uint8)((*(uint32*)&(implParam->u_pulse)) & 0x000000FF);
    SaveFncDataBuffer[1] = (uint8)((*(uint32*)&(implParam->u_pulse) >> 8) & 0x000000FF);
    SaveFncDataBuffer[2] = (uint8)((*(uint32*)&(implParam->u_pulse) >> 16) & 0x000000FF);
    SaveFncDataBuffer[3] = (uint8)((*(uint32*)&(implParam->u_pulse) >> 24) & 0x000000FF);
    SaveFncDataBuffer[4] = (uint8)(implParam->t_pulse & 0x00FF);
    SaveFncDataBuffer[5] = (uint8)((implParam->t_pulse >> 8) & 0x00FF);
    SaveFncDataBuffer[6] = (uint8)(implParam->t_period & 0x00FF);
    SaveFncDataBuffer[7] = (uint8)((implParam->t_period >> 8) & 0x00FF);

    /** - Execute save function of block */
    error = InitPosDetect_Float32_Save(block, SaveFncDataBuffer, 8);

    return (error);
}

/**
 * @brief Converts Mask parameters into Implementation parameters and executes Block Save function.
 *
 * @param[out] block Block Implementation
 * @param[in] maskParam Mask Parameters
 *
 * @return 0 if successful, greater than 0 if conversion failed
 */
uint8 InitPosDetect_Float32_ConvertMP(INITPOSDETECT_FLOAT32 *block, INITPOSDETECT_FLOAT32_MASK_PARAM *maskParam)
{
    uint8 error;
    float_CoT block_ts;
    INITPOSDETECT_FLOAT32_IMPL_PARAM *implParam = (INITPOSDETECT_FLOAT32_IMPL_PARAM*)ImplParamBuffer;

    /** - Get Block Sample Time */
    block_ts = maskParam->ts_fact * getModelSampleTime();
    /** - Convert mask parameters to implementation parameters */
    error = convertM2I(maskParam, implParam, block_ts);
    if (error)
    {
        return (error);
    }

    /** - Prepare implementation parameters for save function */
    SaveFncDataBuffer[0] = (uint8)((*(uint32*)&(implParam->u_pulse)) & 0x000000FF);
    SaveFncDataBuffer[1] = (uint8)((*(uint32*)&(implParam->u_pulse) >> 8) & 0x000000FF);
    SaveFncDataBuffer[2] = (uint8)((*(uint32*)&(implParam->u_pulse) >> 16) & 0x000000FF);
    SaveFncDataBuffer[3] = (uint8)((*(uint32*)&(implParam->u_pulse) >> 24) & 0x000000FF);
    SaveFncDataBuffer[4] = (uint8)(implParam->t_pulse & 0x00FF);
    SaveFncDataBuffer[5] = (uint8)((implParam->t_pulse >> 8) & 0x00FF);
    SaveFncDataBuffer[6] = (uint8)(implParam->t_period & 0x00FF);
    SaveFncDataBuffer[7] = (uint8)((implParam->t_period >> 8) & 0x00FF);

    /** - Execute save function of block */
    error = InitPosDetect_Float32_Save(block, SaveFncDataBuffer, 8);

    return (error);
}

/**
 * @brief Backups current Mask parameters into global Mask parameter backup array.
 *
 * @param[in] maskParam Mask parameters
 *
 * @return Nothing
 */
void InitPosDetect_Float32_BackupMP(INITPOSDETECT_FLOAT32_MASK_PARAM* maskParam)
{
    memcpy(MaskParamBuffer, maskParam, sizeof(*maskParam));
}

/**
 * @brief Restores Mask parameters from global Mask parameter backup array.
 *
 * @param[out] maskParam Mask parameters
 *
 * @return Nothing
 */
void InitPosDetect_Float32_RestoreMP(INITPOSDETECT_FLOAT32_MASK_PARAM* maskParam)
{
    memcpy(maskParam, MaskParamBuffer, sizeof(*maskParam));
}

File: fw/Zsmt_mark/Zsmt_mark/mcc_generated_files/X2CCode/Library/MCHP_ZSMT/Conversion/Java/src/at/lcm/x2c/library/MCHP_ZSMT/ConvFnc_InitPosDetect_Float32.java
/*
 * Copyright (c) 2018, Linz Center of Mechatronics GmbH (LCM) http://www.lcm.at/
 * All rights reserved.
 */
/*
 * This file is part of X2C. http://x2c.lcm.at/
 * $LastChangedRevision: 2108 $
 * $LastChangedDate:: 2020-12-07 19:09:57 +0100#$
 */
// USERCODE-BEGIN:Description                                                                                           
/* Description: */
// USERCODE-END:Description                                                                                             

package at.lcm.x2c.library.mchp_zsmt;

import at.lcm.x2c.core.structure.*;
import at.lcm.bu21.general.dtypes.*;

// USERCODE-BEGIN:Imports                                                                                               
// USERCODE-END:Imports                                                                                                 

@SuppressWarnings("unused")
public class ConvFnc_InitPosDetect_Float32 extends JavaConversionFunction {
    private static final long serialVersionUID = 1L;

    @Override
    public void convertMaskToImplementation() throws Exception {

        // Mask Parameter: U_pulse                                                                                      
        MaskDouble U_pulseMaskVal = 
            (MaskDouble)this.getMaskParameter("U_pulse").getMaskDataType();
        TNumeric U_pulseMaskData = 
            (TNumeric)U_pulseMaskVal.getData();
        // Mask Parameter: T_pulse                                                                                      
        MaskDouble T_pulseMaskVal = 
            (MaskDouble)this.getMaskParameter("T_pulse").getMaskDataType();
        TNumeric T_pulseMaskData = 
            (TNumeric)T_pulseMaskVal.getData();
        // Mask Parameter: T_pause                                                                                      
        MaskDouble T_pauseMaskVal = 
            (MaskDouble)this.getMaskParameter("T_pause").getMaskDataType();
        TNumeric T_pauseMaskData = 
            (TNumeric)T_pauseMaskVal.getData();
        // Mask Parameter: U_max                                                                                        
        MaskDouble U_maxMaskVal = 
            (MaskDouble)this.getMaskParameter("U_max").getMaskDataType();
        TNumeric U_maxMaskData = 
            (TNumeric)U_maxMaskVal.getData();
        // Mask Parameter: ts_fact                                                                                      
        MaskDouble ts_factMaskVal = 
            (MaskDouble)this.getMaskParameter("ts_fact").getMaskDataType();
        TNumeric ts_factMaskData = 
            (TNumeric)ts_factMaskVal.getData();

        // Implementation Parameter: u_pulse                                                                            
        TNumeric u_pulseCtrVal = this.getImplementationParameterData("u_pulse");
        // Implementation Parameter: t_pulse                                                                            
        TNumeric t_pulseCtrVal = this.getImplementationParameterData("t_pulse");
        // Implementation Parameter: t_period                                                                           
        TNumeric t_periodCtrVal = this.getImplementationParameterData("t_period");
        // Implementation Parameter: timer                                                                              
        TNumeric timerCtrVal = this.getImplementationParameterData("timer");
        // Implementation Parameter: dir                                                                                
        TNumeric dirCtrVal = this.getImplementationParameterData("dir");
        // Implementation Parameter: phi                                                                                
        TNumeric phiCtrVal = this.getImplementationParameterData("phi");
        // Implementation Parameter: max_ia_pos                                                                         
        TNumeric max_ia_posCtrVal = this.getImplementationParameterData("max_ia_pos");
        // Implementation Parameter: max_ia_neg                                                                         
        TNumeric max_ia_negCtrVal = this.getImplementationParameterData("max_ia_neg");
        // Implementation Parameter: max_ib_pos                                                                         
        TNumeric max_ib_posCtrVal = this.getImplementationParameterData("max_ib_pos");
        // Implementation Parameter: max_ib_neg                                                                         
        TNumeric max_ib_negCtrVal = this.getImplementationParameterData("max_ib_neg");
        // Implementation Parameter: max_ic_pos                                                                         
        TNumeric max_ic_posCtrVal = this.getImplementationParameterData("max_ic_pos");
        // Implementation Parameter: max_ic_neg                                                                         
        TNumeric max_ic_negCtrVal = this.getImplementationParameterData("max_ic_neg");

// USERCODE-BEGIN:ConvMaskToImplementation                                                                              
        int nPulse, nPause, nPeriod;

        // get mask value
        double ampPhasor = Double.valueOf(U_pulseMaskVal.getValue()); /* get amplitude of voltage phasor */
        double deltaT = Double.valueOf(T_pauseMaskVal.getValue()); /* pause time */
        double ts_fact = Double.valueOf(ts_factMaskVal.getValue()).intValue();
        nPulse = Double.valueOf(T_pulseMaskVal.getValue()).intValue(); /* pulse duration in samples */

        // validation
        if (ts_fact <= 0) {
            throw new Exception("Sample time factor must not be smaller than one!");
        }
        if (nPulse < 1) {
            throw new Exception("Pulse duration must be at least one sample!");
        }
        if (deltaT < 0) {
            throw new Exception("Pause time must not be negative!");
        }

        // calculate implementation parameter value
        /* scaled voltage phasor amplitude */
        u_pulseCtrVal.setReal(0, 0, ampPhasor);

        /* pulse time in updates */
        if (nPulse > 32767) {
            nPulse = 32767;
        }
        t_pulseCtrVal.setReal(0, 0, nPulse);

        /* period in updates */
        nPause = (int) (deltaT / 1000 / (ts_fact * this.getDedicatedBlock().getModel().getSampleTime()));
        nPeriod = 2 * nPulse + nPause;
        if (nPeriod > 32767) {
            nPeriod = 32767;
        }
        t_periodCtrVal.setReal(0, 0, nPeriod);

// USERCODE-END:ConvMaskToImplementation                                                                                
        this.setImplementationParameterData("u_pulse", u_pulseCtrVal);
        this.setImplementationParameterData("t_pulse", t_pulseCtrVal);
        this.setImplementationParameterData("t_period", t_periodCtrVal);
        this.setImplementationParameterData("timer", timerCtrVal);
        this.setImplementationParameterData("dir", dirCtrVal);
        this.setImplementationParameterData("phi", phiCtrVal);
        this.setImplementationParameterData("max_ia_pos", max_ia_posCtrVal);
        this.setImplementationParameterData("max_ia_neg", max_ia_negCtrVal);
        this.setImplementationParameterData("max_ib_pos", max_ib_posCtrVal);
        this.setImplementationParameterData("max_ib_neg", max_ib_negCtrVal);
        this.setImplementationParameterData("max_ic_pos", max_ic_posCtrVal);
        this.setImplementationParameterData("max_ic_neg", max_ic_negCtrVal);
    }

    @Override
    public void convertImplementationToMask() throws Exception {

        // Mask Parameter: U_pulse                                                                                      
        MaskDouble U_pulseMaskVal = 
            (MaskDouble)this.getMaskParameter("U_pulse").getMaskDataType();
        TNumeric U_pulseMaskData = 
            (TNumeric)U_pulseMaskVal.getData();
        // Mask Parameter: T_pulse                                                                                      
        MaskDouble T_pulseMaskVal = 
            (MaskDouble)this.getMaskParameter("T_pulse").getMaskDataType();
        TNumeric T_pulseMaskData = 
            (TNumeric)T_pulseMaskVal.getData();
        // Mask Parameter: T_pause                                                                                      
        MaskDouble T_pauseMaskVal = 
            (MaskDouble)this.getMaskParameter("T_pause").getMaskDataType();
        TNumeric T_pauseMaskData = 
            (TNumeric)T_pauseMaskVal.getData();
        // Mask Parameter: U_max                                                                                        
        MaskDouble U_maxMaskVal = 
            (MaskDouble)this.getMaskParameter("U_max").getMaskDataType();
        TNumeric U_maxMaskData = 
            (TNumeric)U_maxMaskVal.getData();
        // Mask Parameter: ts_fact                                                                                      
        MaskDouble ts_factMaskVal = 
            (MaskDouble)this.getMaskParameter("ts_fact").getMaskDataType();
        TNumeric ts_factMaskData = 
            (TNumeric)ts_factMaskVal.getData();

        // Implementation Parameter: u_pulse                                                                            
        TNumeric u_pulseCtrVal = this.getImplementationParameterData("u_pulse");
        // Implementation Parameter: t_pulse                                                                            
        TNumeric t_pulseCtrVal = this.getImplementationParameterData("t_pulse");
        // Implementation Parameter: t_period                                                                           
        TNumeric t_periodCtrVal = this.getImplementationParameterData("t_period");
        // Implementation Parameter: timer                                                                              
        TNumeric timerCtrVal = this.getImplementationParameterData("timer");
        // Implementation Parameter: dir                                                                                
        TNumeric dirCtrVal = this.getImplementationParameterData("dir");
        // Implementation Parameter: phi                                                                                
        TNumeric phiCtrVal = this.getImplementationParameterData("phi");
        // Implementation Parameter: max_ia_pos                                                                         
        TNumeric max_ia_posCtrVal = this.getImplementationParameterData("max_ia_pos");
        // Implementation Parameter: max_ia_neg                                                                         
        TNumeric max_ia_negCtrVal = this.getImplementationParameterData("max_ia_neg");
        // Implementation Parameter: max_ib_pos                                                                         
        TNumeric max_ib_posCtrVal = this.getImplementationParameterData("max_ib_pos");
        // Implementation Parameter: max_ib_neg                                                                         
        TNumeric max_ib_negCtrVal = this.getImplementationParameterData("max_ib_neg");
        // Implementation Parameter: max_ic_pos                                                                         
        TNumeric max_ic_posCtrVal = this.getImplementationParameterData("max_ic_pos");
        // Implementation Parameter: max_ic_neg                                                                         
        TNumeric max_ic_negCtrVal = this.getImplementationParameterData("max_ic_neg");

// USERCODE-BEGIN:ConvImplementationToMask                                                                              
        if (true) {
            throw new Exception("Implementation-to-Mask conversion not implemented");
        }
// USERCODE-END:ConvImplementationToMask                                                                                
        this.getMaskParameter("U_pulse").setValue(U_pulseMaskData.toFullString());
        this.getMaskParameter("T_pulse").setValue(T_pulseMaskData.toFullString());
        this.getMaskParameter("T_pause").setValue(T_pauseMaskData.toFullString());
        this.getMaskParameter("U_max").setValue(U_maxMaskData.toFullString());
        this.getMaskParameter("ts_fact").setValue(ts_factMaskData.toFullString());
    }
}
